### DISCUSSION TITLE
Is it possible to migrate the S3 storage to a different location

### DISCUSSION THREAD

**Discussion by cwirz** [outline/outline#5286] [General]
I have created a ouline setup that is running now for multiple months and now I would like to migrate the current connected s3 bucket to a selfhosted one. I can move all the files from the current bucket to the selfhosted one with no issues but how do I migrate the content in outline? Is there a guide somewhere?

**Comment by schnz** [comment]
Quite a late response, I just stumbled over this while searching for something else.

I just tried out outline and I am totally new to it, so my knowledge is _very_ limited. But it appears to me, that there doesn't have to be done any migration work, other than simply changing the S3 configuration (bucket name, endpoint, etc.). The database (or at least the `attachments` table) does not store any reference to the bucket, only the relative paths to the files.

Or have you tried that and ran into problems of some kind?

**Reply by axelsimon** [reply]
I've just moved Outline data from a S3 bucket in a locally deployed Minio to a newly created bucket on a remote object storage service (technically, an OpenStack Swift object store, but it makes no difference), using [rclone](https://rclone.org/s3/).

After setting up the two "remotes" (in rclone terms, so 1. the local minio and 2. the remote object store), you can simply do:
`rclone copy local-minio:outline-data remote-object-store:outline-data` and everything just gets copied across.

Then update the `AWS_S3_UPLOAD_BUCKET_URL` and `AWS_S3_UPLOAD_BUCKET_NAME` variables to point to the new object store and bucket name, and you should be good to go.

One downside so far, while all the uploaded files just work, the avatars are broken. It appears Outline is still pointing to their old URL, even with the update of the S3 info. Maybe something that can be tweaked in the database.

**Reply by tommoor** [reply]
> One downside so far, while all the uploaded files just work, the avatars are broken. It appears Outline is still pointing to their old URL, even with the update of the S3 info. Maybe something that can be tweaked in the database.

Yes, the full URL's are stored so you'd need to update the URL in `user.avatarUrl` in the database directly

**Reply by axelsimon** [reply]
Thanks, i was able to update the URLs. In case it can help others:
- drop into the postgres container: `podman exec -it postgres bash`
- connect to the postgres database as the outline user: `psql -U outline`
- select the outline database: `\c outline`
- replace all references to the old URL to the new url in users.avatarUrl: `UPDATE users SET "avatarUrl" = REPLACE("avatarUrl", 'docs.company.com/old-outline-bucket', 'new-S3-store.net/new-outline-bucket');`

Adjust and modify your actual URLs as needed. If you haven't changed S3 bucket name, only the domain names are needed (ie: `'docs.company.com', 'new-S3-store.net'`).

It's not been a success yet however, as i'm still getting 404 errors on my S3 provider's side. It could have to do with outline using POST rather than PUT, or with CORS rules. But at least, it's trying the right URL now :)
